templates/model/decode/decode.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package decode
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"github.com/abcxyz/abc/templates/model"
+	goldentestv1alpha1 "github.com/abcxyz/abc/templates/model/goldentest/v1alpha1"
+	manifestv1alpha1 "github.com/abcxyz/abc/templates/model/manifest/v1alpha1"
+	specv1alpha1 "github.com/abcxyz/abc/templates/model/spec/v1alpha1"
+	specv1beta1 "github.com/abcxyz/abc/templates/model/spec/v1beta1"
+	"golang.org/x/exp/slices"
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	KindTemplate   = "Template"   // the value of the "kind" field in a spec.yaml file
+	KindGoldenTest = "GoldenTest" // ... a test.yaml file
+	KindManifest   = "Manifest"   // ... a manifest.yaml file
+)
+
+type apiVersionDef struct {
+	apiVersion string
+
+	// Map keys are the "kind" values found in the YAML files.
+	kinds map[string]model.ValidatorUpgrader
+}
+
+// This list is in chronological order of API release (oldest to newest). To
+// remove support for an api_version, delete from the head of the list. To add a
+// new api_version, append to the end of the list. See
+// templates/model/README.md for detailed instructions on creating a new
+// api_version.
+
+// Typically, you'll only be changing one of the model types
+// (spec/manifest/goldentest) when introducing a new api_version. In that case
+// just copy-paste the old model types from the previous api_version.
+var apiVersions = []apiVersionDef{
+	{
+		apiVersion: "cli.abcxyz.dev/v1alpha1",
+		kinds: map[string]model.ValidatorUpgrader{
+			KindTemplate:   &specv1alpha1.Spec{},
+			KindGoldenTest: &goldentestv1alpha1.Test{},
+			KindManifest:   &manifestv1alpha1.Manifest{},
+		},
+	},
+	{
+		apiVersion: "cli.abcxyz.dev/v1beta1",
+		kinds: map[string]model.ValidatorUpgrader{
+			KindTemplate:   &specv1beta1.Spec{},
+			KindGoldenTest: &goldentestv1alpha1.Test{},
+			KindManifest:   &manifestv1alpha1.Manifest{},
```

Comment:
Maybe these should be separate structs, in different files and we only add individual api version changes? This would make it clearer about which versions are supported by which `Kinda`. This would also make it "easier" to automate later by targeting a specific file per kind.

Reply:
Spoke offline, the current code seems like the best solution currently.